src/server/viper.ts

```diff
@@ -0,0 +1,271 @@
+    //     },
+    // );
+
+    connection.onNotification(
+        new rpc.NotificationType<{ uri: string }>('VerificationNotStarted'),
+        () => {
+            console.log('(Verification not started)');
+        },
+    );
+
+    connection.onNotification(
+        new rpc.NotificationType<{ data: string; logLevel: number }>('Log'),
+        ({ data }) => {
+            console.log(data);
+        },
+    );
+
+    const showMessageTypes = [
+        'warnings_during_parsing',
+        // 'configuration_confirmation',
+        'ast_construction_result',
+    ];
+    connection.onNotification(
+        new rpc.NotificationType<{
+            msgType: string;
+            msg: string;
+            logLevel: number;
+        }>('UnhandledViperServerMessageType'),
+        ({ msgType, msg }) => {
+            if (showMessageTypes.includes(msgType)) {
+                console.log(msg);
+            } else {
+                console.log(`[${msgType}]`);
+            }
+        },
+    );
+} catch (err) {
+    console.error(`Error while initializing Viper LSP: ${err}`);
+}
+
+// Temporary range input format for `moc.js`
+type FlattenedRange = [number, number, number, number];
+
+// Viper -> Motoko source map cache
+interface ViperResult {
+    source: string;
+    lookup(motokoPath: string, pos: FlattenedRange): Range;
+    diagnostics: Diagnostic[];
+}
+const mocViperCache = new Map<string, ViperResult>();
+
+function getMotokoSourceRange(
+    motokoPath: string,
+    { start: { line: a, character: b }, end: { line: c, character: d } }: Range,
+): Range | undefined {
+    const result = mocViperCache.get(motokoPath);
+    if (!result) {
+        return;
+    }
+    return result.lookup(motokoPath, [a, b, c, d]); // TODO: directly pass range
+}
+
+export function getViperUri(motokoUri: string) {
+    // Ensure a `.vpr` extension
+    return `${motokoUri /* .replace(/\.mo$/, '') */}.vpr`;
+}
+
+export function getMotokoUri(viperUri: string) {
+    // Reversible from `getViperUri()`
+    return viperUri.replace(/\.vpr$/, '');
+}
+
+let verifyTimeout: ReturnType<typeof setTimeout>;
+export function compileViper(motokoUri: string): Diagnostic[] | undefined {
+    const viperUri = getViperUri(motokoUri);
+    const viperFile = resolveFilePath(viperUri);
+    const motokoPath = resolveVirtualPath(motokoUri);
+    let diagnostics: Diagnostic[] | undefined;
+    try {
+        const result = mo.compiler.viper([motokoPath]);
+        if (result?.diagnostics) {
+            diagnostics = result.diagnostics;
+        }
+        if (result?.code) {
+            const { viper: source, lookup } = result.code;
+            mocViperCache.set(motokoPath, {
+                source,
+                lookup,
+                diagnostics: result.diagnostics || [],
+            });
+            writeFileSync(viperFile, source, 'utf8');
+
+            // Debounce verification
+            clearTimeout(verifyTimeout);
+            verifyTimeout = setTimeout(() => {
```

Comment:
It was necessary to add a debounce (using the confusingly named JS `setTimeout` API) to avoid hitting a limit to the number of active verifications in the Viper LS. This logic waits to start verification until 500ms after a user finishes typing (rather than on every keystroke), which was more than enough to resolve this issue.